python/python/lance/sampler.py

```diff
@@ -343,27 +358,100 @@ def from_torch(randomize: bool = False, seed: int = 0) -> ShardedBatchSampler:
         world_size = torch.distributed.get_world_size()
         return ShardedBatchSampler(rank, world_size, randomize=randomize, seed=seed)
 
-    def __call__(
+    # Performs a filtered scan of the dataset and then throws away all but the Nth
+    # rows (where N is the world size)
+    def _shard_scan(
         self,
         dataset: lance.LanceDataset,
-        *args,
-        batch_size: int = 128,
-        columns: Optional[List[str]] = None,
-        filter: Optional[str] = None,
-        batch_readahead: int = 16,
-        with_row_id: Optional[bool] = None,
-        **kwargs,
+        batch_size: int,
+        columns: Optional[List[str]],
+        batch_readahead: int,
+        filter: str,
+    ) -> Generator[lance.RecordBatch, None, None]:
+        accumulated_batches = []
+        rows_accumulated = 0
+        rows_to_skip = self._rank
+        for batch in dataset.scanner(
+            columns=columns,
+            batch_readahead=batch_readahead,
+            filter=filter,
+            scan_in_order=True,
+        ).to_batches():
+            batch = batch.slice(rows_to_skip, batch.num_rows - rows_to_skip)
+            # Take every Nth row
+            indices = list(range(0, batch.num_rows, self._world_size))
+            rows_to_skip = (self._world_size - (batch.num_rows % self._world_size)) % 5
```

Comment:
What is this magic number `5` here?

Reply:
Yikes.  Good catch.  That's what I get for focusing too much on a unit test.  That was supposed to be `self._world_size`.  I've fixed it.